test(supervisors): fix occasional assertion failures and hangs

vvanglro opened the pull request:
<!-- Thanks for contributing to Uvicorn! 💚
Given this is a project maintained by volunteers, please read this template to not waste your time, or ours! 😁 -->

# Summary
It's not entirely clear why the assertion fails here. `assert not process.is_alive()` 

Make sure that test resources can be cleaned up anyway.

<!-- Write a small summary about what is happening here. -->

# Checklist

- [x] I understand that this PR may be closed in case there was no previous discussion. (This doesn't apply to typos!)
- [x] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [x] I've updated the documentation accordingly.

vvanglro committed:
test(supervisors): ensure cleanup in test_multiprocess by using 'finally'

Ensure proper cleanup in 'test_multiprocess' by wrapping the supervisor shutdown
logic in a 'finally' block. This guarantees that the supervisor is properly terminated
and all processes are joined, avoiding potential resource leaks or state residuebetween test runs.

vvanglro committed:
test(supervisor): add timeout to multiprocess test subprocess

A10-second timeout has been added to the subprocess call within the test_multiprocess.py
to ensure it does not hang indefinitely in case of failures or slow systems.

vvanglro committed:
try to solve

vvanglro committed:
try to solve

vvanglro committed:
Merge branch 'master' into fix/test_hang

vvanglro committed:
feat: add retry for unstable tests

vvanglro commented on tests/supervisors/test_multiprocess.py:
```diff
@@ -75,6 +76,7 @@ def test_multiprocess_run() -> None:
     supervisor.join_all()
 
 
+@pytest.mark.flaky(reruns=3, reruns_delay=2)
```
@Kludex For unstable test cases, how about add a retry?

Kludex replied:
No. We should make them stable.

vvanglro replied:
This feels a bit difficult, because the service is started with a thread. And the run method involves multiple waits.

https://github.com/encode/uvicorn/blob/master/uvicorn/supervisors/multiprocess.py#L151-L153

Switching execution between threads, as well as killed processes, all affect test stability and assertions.

vvanglro committed:
test(supervisor): remove retry, ensure process close.

Kludex commented on tests/supervisors/test_multiprocess.py:
```diff
@@ -83,15 +84,18 @@ def test_multiprocess_health_check() -> None:
     supervisor = Multiprocess(config, target=run, sockets=[])
     threading.Thread(target=supervisor.run, daemon=True).start()
-    time.sleep(1)
+    time.sleep(1)  # ensure server is up
```
```suggestion
    # Ensure server is up.
    time.sleep(1)
```

vvanglro committed:
Merge branch 'master' into fix/test_hang

Kludex commented on tests/supervisors/test_multiprocess.py:
```diff
@@ -34,6 +34,7 @@ def new_function():
                 "-c",
                 f"from {module} import {name}; {name}.__wrapped__()",
             ],
+            timeout=10,
```
Why do we need the timeout?

vvanglro replied:
Oh, this was added earlier when I was investigating why it was hanging. No need for it now, I'll remove it later

vvanglro committed:
test case

vvanglro committed:
Merge branch 'master' into fix/test_hang

vvanglro committed:
Merge branch 'master' into fix/test_hang

vvanglro committed:
Merge branch 'master' into fix/test_hang

Kludex committed:
Merge branch 'master' into fix/test_hang

vvanglro committed:
test(supervisors): improve test_multiprocess reliability

- Replace fixed sleep with retry mechanism for process status checks- Enhance test stability by allowing for variable process startup times

vvanglro committed:
test(supervisors): improve test_multiprocess reliability

vvanglro committed:
Merge branch 'master' into fix/test_hang

vvanglro committed:
test(supervisors): improve test_multiprocess reliability

vvanglro committed:
Merge branch 'master' into fix/test_hang

vvanglro committed:
Release gil to let the supervisor thread switch execution

vvanglro committed:
format

vvanglro committed:
Merge branch 'master' into fix/test_hang

Kludex committed:
Merge branch 'master' into fix/test_hang

Kludex reviewed (approved):
Sorry the delay!

vvanglro committed:
Merge branch 'master' into fix/test_hang

vvanglro committed:
test(supervisors): improve test_multiprocess reliability

- Add retry mechanism for unstable test_multiprocess_health_check- Ensure server is running before killing the process
- Add assertion to check process is alive before killing it

vvanglro committed:
Merge branch 'master' into fix/test_hang

vvanglro committed:
Merge branch 'master' into fix/test_hang

vvanglro committed:
test(supervisors): implement a generic retry mechanism for tests

- Add a `with_retry` decorator in tests/utils.py to handle retries
- Apply the decorator to the `test_multiprocess_health_check` function
- Remove the ad-hoc retry logic specific to `test_multiprocess_health_check`

vvanglro committed:
test: add pragma nocover to retry decorator

- Add pragma nocover comment to async_wrapper and sync_wrapper functions- This change excludes the decorator logic from code coverage metrics

vvanglro committed:
lint

vvanglro committed:
test: add pragma nocover to async_wrapper return statement

- Add pragma nocover comment to exclude async_wrapper return statement from coverage
- This change helps to avoid unnecessary test coverage metrics
